server/src/main/java/org/elasticsearch/plugins/PluginDescriptor.java

```diff
@@ -250,6 +250,31 @@ public static PluginDescriptor readFromProperties(final Path pluginDir) throws I
         return descriptor;
     }
 
+    /**
+     * Reads the internal descriptor for a classic plugin.
+     *
+     * @param stream the InputStream from which to read the plugin data
+     * @return the plugin info
+     * @throws IOException if an I/O exception occurred reading the plugin descriptor
+     */
+    public static PluginDescriptor readInternalDescriptor(InputStream stream) throws IOException {
```

Comment:
nit: readFromStream to match readFromProperties above?

Reply:
I think this does less than `readFromProperties`: it does not support stable plugins.

I can go with `readInternalDescriptorFromStream` if that's better.